Port `sum.dim_IntList` kernel to structured kernels.

ysiraichi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #62711 Port `norm` kernel to structured kernels.
* #62024 Port `prod` kernel to structured kernels.
* #61643 Port `mean` kernel to structured kernels.
* **#61642 Port `sum.dim_IntList` kernel to structured kernels.**

Tracking issue: #55070

Differential Revision: [D29783865](https://our.internmc.facebook.com/intern/diff/D29783865)

ysiraichi committed:
Port `sum.dim_IntList` kernel to structured kernels.

Tracking issue: #55070

[ghstack-poisoned]

ysiraichi committed:
Update on "Port `sum.dim_IntList` kernel to structured kernels."


Tracking issue: #55070

[ghstack-poisoned]

ysiraichi committed:
Update on "Port `sum.dim_IntList` kernel to structured kernels."


Tracking issue: #55070

[ghstack-poisoned]

ysiraichi committed:
Update on "Port `sum.dim_IntList` kernel to structured kernels."


Tracking issue: #55070

[ghstack-poisoned]

ysiraichi commented:
In one of the CI failures, I'm getting on the `empty_cuda` kernel. Although it is related (the `empty_cuda` call comes from `sum::meta` call), I don't think the error is caused by the kernel, specifically. Any thoughts?

```
C++ exception with description "0 <= device && static_cast<size_t>(device) < device_allocator.size()INTERNAL ASSERT FAILED at "/var/lib/jenkins/workspace/c10/cuda/CUDACachingAllocator.cpp":1352, please report a bug to PyTorch. Allocator not initialized for device 0: did you call init?
```

ezyang commented on aten/src/ATen/native/ReduceOps.cpp:
```diff
@@ -116,6 +134,25 @@ TORCH_META_FUNC(argmin)
+  const auto& result = maybe_get_output();
+
+  if (result.defined()) {
+    dtype = opt_dtype.value_or(result.scalar_type());
```
This made me raise an eyebrow but it is compatible with the old behavior of `get_dtype_from_result`

```
inline ScalarType get_dtype_from_result(Tensor& result, optional<ScalarType> dtype) {
  TORCH_CHECK(result.defined(), "Cannot create a new tensor inside a reduction op. You likely tried to call an operator with an out argument but the out argument was an undefined tensor.");
  if (dtype.has_value()) {
    return dtype.value();
  } else {
    return result.scalar_type();
  }
}

```

ezyang commented on aten/src/ATen/native/ReduceOps.cpp:
```diff
@@ -116,6 +134,25 @@ TORCH_META_FUNC(argmin)
+  }
+
+  DimVector dims(dim);
+  maybe_wrap_dims(dims, self.dim());
```
I expected something corresponding to this code to be deleted in the other part of this PR, wondering why we can't delete it yet. Is this because this is another composite-y kernel? Kind of looks like it...

ysiraichi replied:
Not sure if I got your question. Are you asking what's the relation between this PR and the `mean` one? If so, I guess this is my bad. After having ported them both, I realized it didn't make sense to group them into helper functions (they have the same logic, but don't necessarily have to).

ezyang commented on aten/src/ATen/native/ReduceOps.cpp:
```diff
@@ -875,38 +912,22 @@ inline ScalarType get_dtype_from_result(Tensor& result, optional<ScalarType> dty
+ bool keepdim,
+ optional<ScalarType> opt_dtype,
+ const Tensor& result) {
+  auto iter = meta::make_reduction_from_out_ty(self, result, dim, keepdim, result.scalar_type());
```
Remind me again why `sum` doesn't use TensorIteratorBase directly?

ezyang replied:
oh is this the thing where reductions are weird

ysiraichi replied:
Yes! Basically, `TensorIteratorBase` expects the to-be-reduced dimensions to have size 1 and stride 0, so that it can reorder the dimensions. In the current state, I think there's no way to do that without changing the result's shape and properties.

ezyang commented:
Looking over the PR, I suspect you were just unlucky enough to hit a latent bug. Checking...

ezyang commented:
@ysiraichi I think this is a latent bug triggered by structured kernels. Try this patch:

```
diff --git a/tools/codegen/dest/register_dispatch_key.py b/tools/codegen/dest/register_dispatch_key.py
index fc2af8fc13..a943f51ba5 100644
--- a/tools/codegen/dest/register_dispatch_key.py
+++ b/tools/codegen/dest/register_dispatch_key.py
@@ -356,10 +356,12 @@ if (strides.empty()) {
             else:
                 expanded_topts = "optTypeMetaToScalarType(options.dtype_opt()), options.layout_opt(), " \
                     "options.device_opt(), options.pinned_memory_opt()"
+                empty_init = ""
                 if self.backend_index.dispatch_key == DispatchKey.CPU:
                     empty_impl = "at::native::empty_cpu"
                     empty_strided_impl = "at::native::empty_strided_cpu"
                 elif self.backend_index.dispatch_key == DispatchKey.CUDA:
+                    empty_init = "globalContext().lazyInitCUDA();"
                     empty_impl = "at::native::empty_cuda"
                     empty_strided_impl = "at::native::empty_strided_cuda"
                 elif self.backend_index.dispatch_key == DispatchKey.CompositeExplicitAutograd:
@@ -368,6 +370,7 @@ if (strides.empty()) {
                 else:
                     raise AssertionError("unsupported dispatch key")
                 return f"""{maybe_set_guard_line}
+{empty_init}
 if (strides.empty()) {{
     outputs_[output_idx] = {empty_impl}(sizes, {expanded_topts}, options.memory_format_opt());
 }} else {{

```

ezyang commented:
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29783865).

ysiraichi commented:
@ezyang Yep. That solved the issue!

ysiraichi committed:
Update on "Port `sum.dim_IntList` kernel to structured kernels."


Tracking issue: #55070

Differential Revision: [D29783865](https://our.internmc.facebook.com/intern/diff/D29783865)

[ghstack-poisoned]

ezyang commented:
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29783865).

ezyang committed:
Update on "Port `sum.dim_IntList` kernel to structured kernels."


Tracking issue: #55070

Differential Revision: [D29783865](https://our.internmc.facebook.com/intern/diff/D29783865)

[ghstack-poisoned]

ezyang commented:
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29783865).

ysiraichi committed:
Update on "Port `sum.dim_IntList` kernel to structured kernels."


Tracking issue: #55070

Differential Revision: [D29783865](https://our.internmc.facebook.com/intern/diff/D29783865)

[ghstack-poisoned]

ysiraichi commented:
Rebased.

ysiraichi committed:
Update on "Port `sum.dim_IntList` kernel to structured kernels."


Tracking issue: #55070

Differential Revision: [D29783865](https://our.internmc.facebook.com/intern/diff/D29783865)

[ghstack-poisoned]

ysiraichi committed:
Update on "Port `sum.dim_IntList` kernel to structured kernels."


Tracking issue: #55070

Differential Revision: [D29783865](https://our.internmc.facebook.com/intern/diff/D29783865)

[ghstack-poisoned]

ezyang commented:
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29783865).

ezyang committed:
Update on "Port `sum.dim_IntList` kernel to structured kernels."


Tracking issue: #55070

Differential Revision: [D29783865](https://our.internmc.facebook.com/intern/diff/D29783865)

[ghstack-poisoned]

ezyang commented:
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29783865).

ysiraichi committed:
Update on "Port `sum.dim_IntList` kernel to structured kernels."


Tracking issue: #55070

Differential Revision: [D29783865](https://our.internmc.facebook.com/intern/diff/D29783865)

[ghstack-poisoned]

ezyang commented:
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29783865).

ezyang commented:
sorry about this being so slow to land, hit some mobile related hiccups
